### DISCUSSION TITLE
Things to include in paper

### DISCUSSION THREAD

**Discussion by JoelMathewC** [bfGraph/STGraph#60] [Ideas]
Conference: https://hipc.org/call-for-papers/

1. Our system is for offline full-batch training of discrete time dynamic graph based GNNs and the general extension for static temporal graphs.
4. Do not use any AI generated text (if used then attribute). As per IEEE guidelines, the use of artificial intelligence (AI)–generated text in an article needs to be disclosed in the acknowledgements section of any submitted paper. The sections of the paper that use AI-generated text should have a citation to the AI system used to generate the text. For more information, [please click here](https://journals.ieeeauthorcenter.ieee.org/become-an-ieee-journal-author/publishing-ethics/guidelines-and-policies/submission-and-peer-review-policies/#ai-generated-text).
5. The IEEE conference style templates for MS Word and LaTeX provided by IEEE eXpress Conference Publishing are available for download. See the latest versions [here](https://www.ieee.org/conferences/publishing/templates.html). Electronic submissions must be in the form of a readable PDF file.
6. Authors are highly encouraged to list the key contributions of their paper, for example in a separate paragraph in the introduction of the paper. The review process is “double-blind” (i.e., authors should not include their names or affiliations in the manuscript and all references to the previous works of the authors should be mentioned in third person). 

## Things to note

1. Abstract should not be more than 300 words
2. Full papers may not exceed ten (10) single-spaced double-column pages using 10-point size font on 8.5×11 inch pages (IEEE conference style), including figures, tables, and references.
3. At the time of publishing, authors of accepted papers may buy upto two pages in addition to the page limits specified above for their manuscript.

**Comment by nithinmanoj10** [comment]
# Diagrams for Seastar Paper

- [ ] Diagram showing the structure of old seastar and new seastar
- [ ]  Diagram showing state stack (static temporal graphs) [Edit the diagram such that the graph is static]
- [ ]  Diagram showing state stack + timestamp stack (dynamic temporal graphs)
- [ ]  Seastar graph abstraction [Refer UML specifics when making]
- [ ] Count sort diagram 
- [ ] Backend class diagram

**Comment by nithinmanoj10** [comment]
# Inspiration

Ideas from other research papers

1. [Graphical Excellence in Scientific Presentations and Papers](https://www3.nd.edu/~pkamat/pdf/graphs.pdf)

**Comment by nithinmanoj10** [comment]
# Things added to Seastar from our end

1. Decoupling Seastar from DGL-Hack
2. Provided abstraction (factory class design principle) which enables Seastar to be backend agnostic (kernel wrapper, val and op). 
3. CUDA Python Migrate
4. Seastar Graph Abstraction - StaticGraph, DynamicGraph, PCSRGraph, GPMAGraph and NaiveGraph. PyBind11 was used to create a python interface for these graph implementations written in CUDA.
5. GPMA optimization with pinned memory and count sort for reversing. Wrote a custom kernel for labelling.
6. Introduced the state-stack to keep track of information of tensors for backward propagation of temporal graphs. Only pushes the required tensors at a particular timestamp to use lesser memory.
7. Provided a seastar dataset loader that integrates well into the Seastar ecosystem. 
8. Separate kernel template for GPMA.
9. Timestamp stack for dynamic graphs was introduced so that kernel has knowledge about the current timestamp it is running.
10. Exposing GCN, GAT and T-GCN APIs
11. Modified GPMA code to be more efficient by using in-place radix sort (in this fuction cub_sort_key_value).
12. Added support for code generation of subtraction operator.
13. **Productionizing Seastar** - Docs, testing, tutorials, re-organizing the codebase
14. **Benchmarking** - Use of synthetic dataset, and explanation for why better performance

**Comment by JoelMathewC** [comment]
## Benchmarking Graphs

- [ ] Graph showing performance of Naive, PCSR, GPMA and PyG-T based on feature size

**Comment by JoelMathewC** [comment]
## Mention

1. Mention version of PyG-T
2. Mention why using the PyG-T structure to build TGCN was used
3. Talk about maintainability of DGL Hack and struggle to run
4. Look into how PyG improved their memory from since the paper because the outputs seem to be a lot better now, or seastar has gotten a lot slower. 
5. Benchmark with PyG-T (preloaded) and on-demand versions.
6. Mention we used temporal edges for the SNAP datasets.
7. Look into if the extending pytorch `save_backward_tensors` could have been used instead of the state stack
8. GPMA updates on the GPU and removed sorting since updates can be preprocessed to be sorted
9. Talk about caching in dynamic graphs to allow for `backprop-every`
10. Look into if DGL can support dynamic graphs
11. Naive has massive preprocessing overhead that makes it unscalable. Thats why gpma and pcsr

**Comment by JoelMathewC** [comment]
Closing this out since the paper is done.
